pkg/kubelet/kubelet_volumes.go

```diff
@@ -147,55 +200,47 @@ func (kl *Kubelet) cleanupOrphanedPodDirs(pods []*v1.Pod, runningPods []*kubecon
 			continue
 		}
 
-		allVolumesCleanedUp := true
-
-		// If there are still volume directories, attempt to rmdir them
-		volumePaths, err := kl.getPodVolumePathListFromDisk(uid)
-		if err != nil {
-			orphanVolumeErrors = append(orphanVolumeErrors, fmt.Errorf("orphaned pod %q found, but error %v occurred during reading volume dir from disk", uid, err))
+		// Attempt to remove the pod volumes directory and its subdirs
+		podVolumeErrors := kl.removeOrphanedPodVolumeDirs(uid)
+		if len(podVolumeErrors) > 0 {
+			orphanVolumeErrors = append(orphanVolumeErrors, podVolumeErrors...)
+			// Not all volumes were removed, so don't clean up the pod directory yet. It is likely
+			// that there are still mountpoints or files left which could cause removal of the pod
+			// directory to fail below.
+			// Errors for all removal operations have already been recorded, so don't add another
+			// one here.
 			continue
 		}
-		if len(volumePaths) > 0 {
-			for _, volumePath := range volumePaths {
-				if err := syscall.Rmdir(volumePath); err != nil {
-					orphanVolumeErrors = append(orphanVolumeErrors, fmt.Errorf("orphaned pod %q found, but failed to rmdir() volume at path %v: %v", uid, volumePath, err))
-					allVolumesCleanedUp = false
-				} else {
-					klog.InfoS("Cleaned up orphaned volume from pod", "podUID", uid, "path", volumePath)
-				}
-			}
-		}
 
-		// If there are any volume-subpaths, attempt to rmdir them
-		subpathVolumePaths, err := kl.getPodVolumeSubpathListFromDisk(uid)
+		// Call RemoveAllOneFilesystem for remaining subdirs under the pod directory
+		podDir := kl.getPodDir(uid)
+		podSubdirs, err := ioutil.ReadDir(podDir)
 		if err != nil {
-			orphanVolumeErrors = append(orphanVolumeErrors, fmt.Errorf("orphaned pod %q found, but error %v occurred during reading of volume-subpaths dir from disk", uid, err))
+			klog.ErrorS(err, "Could not read directory", "path", podDir)
+			orphanRemovalErrors = append(orphanRemovalErrors, fmt.Errorf("orphaned pod %q found, but error %v occurred during reading the pod dir from disk", uid, err))
 			continue
 		}
-		if len(subpathVolumePaths) > 0 {
-			for _, subpathVolumePath := range subpathVolumePaths {
-				if err := syscall.Rmdir(subpathVolumePath); err != nil {
-					orphanVolumeErrors = append(orphanVolumeErrors, fmt.Errorf("orphaned pod %q found, but failed to rmdir() subpath at path %v: %v", uid, subpathVolumePath, err))
-					allVolumesCleanedUp = false
-				} else {
-					klog.InfoS("Cleaned up orphaned volume subpath from pod", "podUID", uid, "path", subpathVolumePath)
-				}
+		for _, podSubdir := range podSubdirs {
+			podSubdirName := podSubdir.Name()
+			podSubdirPath := filepath.Join(podDir, podSubdirName)
+			// Never attempt RemoveAllOneFilesystem on the volumes directory,
+			// as this could lead to data loss in some situations. The volumes
+			// directory should have been removed by removeOrphanedPodVolumeDirs.
+			if podSubdirName == "volumes" {
```

Comment:
basically if this happens - it is almost guaranteed that pod cleanup will fail because `syscall.Rmdir(podDir)` below will fail. In that case - shouldn't we log this as error?

Reply:
Yes, good point, I changed this to an error log message. This condition "shouldn't" happen, at least in the absence of any race conditions. If we hit an error in `removeOrphanedPodVolumeDirs` then we'll stop before entering this podSubdir loop. But if the volumes directory got removed successfully and then re-created by another process, we could hit this line, and it would be fair to treat that as an error since the final rmdir call will fail.

Reply:
Well, after I pushed that change, `pull-kubernetes-verify` failed with:
```
kubelet_volumes.go:230:5: unstructured logging function "Error" should not be used
```
The only reason I tried to use `Error` instead of `ErrorS` is because we do not have a specific `err` to pass along to the log in this case. `ErrorS` requires an error as the first argument, while `InfoS` does not. I'm going to revert this back to an `InfoS` call for now, but let me know if you have a better idea. The goal of this log statement is just to leave behind a breadcrumb in case the `volumes` directory somehow gets recreated after `removeOrphanedPodVolumeDirs` returns success.

Reply:
I have not looked into structured logging proposal too deeply but traditionally monitoring systems parse error level logs differently than info logs (I don't know why we don't have `WarningS`) and hence I would prefer to log it as error even if we have to instantiate an error type via `errors.New()`.

Reply:
Thanks, I think Errorf is what I was looking for, and I see some other components using it so hopefully the verify check will pass this time. If not, I'll follow pohly's suggestion.

Reply:
> I don't know why we don't have WarningS

https://github.com/go-logr/logr#why-not-named-levels-like-infowarningerror

Reply:
Error condition looks good. See also https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#change-log-functions-to-structured-equivalent for there not being "WarningS" equivalents.